src/main/java/net/dries007/tfc/objects/blocks/BlockFluidTFC.java

```diff
@@ -0,0 +1,424 @@
+                            minMeta = neighborMeta;
+                        }
+                        else if (neighborMeta == minMeta)
+                        {
+                            if (neighborBlock.getDensity() > blockType.getDensity() ||
+                                (neighborBlock == this && neighborBlock.getDensity() >= blockType.getDensity()))
+                            {
+                                blockType = neighborBlock;
+                            }
+                        }
+                    }
+                }
+            }
+
+            if (minMeta + 1 < currentMeta && blockType != this)
+            {
+                world.setBlockState(pos, blockType.getDefaultState().withProperty(LEVEL, currentMeta), 3);
+            }
+        }
+    }
+
+    @Override
+    public boolean canDisplace(IBlockAccess world, BlockPos pos)
+    {
+        IBlockState state = world.getBlockState(pos);
+        Block block = state.getBlock();
+
+        if (block.isAir(state, world, pos))
+        {
+            return true;
+        }
+
+        if (block == this)
+        {
+            return false;
+        }
+
+        if (displacements.containsKey(block))
+        {
+            return displacements.get(block);
+        }
+
+        Material material = state.getMaterial();
+        if (material.blocksMovement() || material == Material.PORTAL || material == Material.STRUCTURE_VOID)
+        {
+            return false;
+        }
+
+        // this is where it differs from the source:
+
+        if (block instanceof BlockFluidTFC)
+        {
+            return (state.getValue(LEVEL) != 0);
+        }
+
+        int density = getDensity(world, pos);
+        if (density == Integer.MAX_VALUE)
+        {
+            return true;
+        }
+
+        return this.density > density;
+    }
+
+    @Override
+    public float getFluidHeightForRender(IBlockAccess world, BlockPos adjPos, @Nonnull IBlockState upState)
+    {
+        IBlockState adjState = world.getBlockState(adjPos);
+
+        // any adjacent above matching liquids merge to 1
+        if (isMergeableFluid(upState))
+        {
+            return 1;
+        }
+
+        // adjacent mergeable liquids
+        if (isMergeableFluid(adjState))
+        {
+            Block adjBlock = adjState.getBlock();
+            if (adjBlock == this || adjBlock instanceof BlockLiquid)
+                return super.getFluidHeightForRender(world, adjPos, upState);
+            else
+                return ((BlockFluidBase) adjBlock).getFluidHeightForRender(world, adjPos, upState);
+        }
+
+        // adjacent solid
+        if (adjState.getMaterial().isSolid())
+            return -1;
+
+        // adjacent air or non-mergeable liquids
+        return 0;
+    }
+
+    protected boolean isMergeableFluid(@Nonnull IBlockState blockstate)
+    {
+        return (blockstate.getMaterial() == getDefaultState().getMaterial()) && (blockstate.getMaterial().isLiquid());
```

Comment:
I think this should do a `blockstate.getBlock() instanceof BlockFluidTFC` instead of comparing by material.
 - I don't see a particular reason why water can't merge with alchohol - It's certainly not obvious to the player why not.
 - We don't want to override fluid blocks that aren't TFC ones.
If there's something I'm missing here let me know.